### PR TITLE
replay/diff refactoring

### PR DESCRIPTION
**Pull request by apahim**
v1:
- Move avocado.core.replay to avocado.core.jobdata
- Make avocado.core.jobdata more "pylint friendly".
- Rename record diretory from `replay` to `jobdata`.
- Use `replay` directory name as a fallback location to retrieve data from.
- Adjust tests accordingly.
- Move job results probe logic from `avocado.plugins.diff` to `avocado.core.jobdata` so replay can also benefit from the improved  results directory probe.

### CODE REVIEW

File: avocado/core/job.py [hunk 1]
```diff
@@ -42,7 +42,7 @@
 from . import multiplexer
```
Comment (ldoktor) [inline avocado/core/job.py]: Well based on the commit message and diff between the two files it looks like you changed several things at once, which is not really nice. You should fix style, in next commit add improvements and in other commit move things around, otherwise it's hard to make sure there were no typos/unintended changes.

File: avocado/core/jobdata.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+
+from . import exit_codes
+from .settings import settings
+from ..utils.path import init_dir
+
+
+JOB_DATA_DIR = 'jobdata'
+JOB_DATA_FALLBACK_DIR = 'replay'
+CONFIG_FILENAME = 'config'
```
Comment (ldoktor) [inline avocado/core/jobdata.py]: I don't think the config/urls/multiplex/... should be defined as variables. They basically copy the variables, making it clear even without definition.

File: avocado/core/job.py (hunk 1 above)
Comment (apahim) [inline avocado/core/job.py]: Indeed. Sending a v2 with the file move in a separate commit.

File: avocado/core/jobdata.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+    if urls:
+        with open(path_urls, 'w') as urls_file:
+            urls_file.write('%s' % urls)
+
+    with open(path_cfg, 'w') as config_file:
+        settings.config.write(config_file)
+
+    with open(path_mux, 'w') as mux_file:
+        pickle.dump(mux, mux_file, pickle.HIGHEST_PROTOCOL)
```
Comment (ldoktor) [inline avocado/core/jobdata.py]: Question: do we need the HIGHEST protocol? We might want to move the results from one machine to another, which might not use the same pickle version...

File: avocado/core/jobdata.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+    with open(recorded_pwd, 'r') as pwd_file:
+        return pwd_file.read()
+
+
+def retrieve_urls(resultsdir):
+    """
+    Retrieves the job urls from the results directory.
+    """
+    recorded_urls = os.path.join(resultsdir, JOB_DATA_DIR, URLS_FILENAME)
```
Comment (ldoktor) [inline avocado/core/jobdata.py]: I can see a pattern here, you should implement generic `_retrieve` function and then only use it in the `retrieve_*` commands.

File: avocado/core/jobdata.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        return ast.literal_eval(cmdline_file.read())
+
+
+def get_resultsdir(logdir, jobid):
+    """
+    Gets the job results directory using a Job ID.
+    """
+    if jobid == 'latest':
+        try:
+            actual_dir = os.readlink(os.path.join(logdir, 'latest'))
```
Comment (ldoktor) [inline avocado/core/jobdata.py]: This works now, but we might some day use non-relative links to latest and this will be suddenly broken. It should be `os.realpath(os.path.join(logdir, 'latest'))` instead. Also I'm not sure why do you do this as `replay` should work with any directory containing `id` so this could simply just check whether the directory exists (and we'd get support for other existing directories inside the logdir).

Anyway my concern is only the first part, which is that this is potential dangerous and can return incorrect path.

File: avocado/core/jobdata.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    idfile_pattern = os.path.join(logdir, 'job-*-%s' % short_jobid, 'id')
+    for id_file in glob.glob(idfile_pattern):
+        if get_id(id_file, jobid) is not None:
+            match_file = id_file
+            matches += 1
+            if matches > 1:
+                from logging import getLogger
+                getLogger("avocado.app").error("hash '%s' is not unique "
+                                               "enough", jobid)
+                sys.exit(exit_codes.AVOCADO_JOB_FAIL)
```
Comment (ldoktor) [inline avocado/core/jobdata.py]: We should not `sys.exit` directly from inside the library. It should raise an exception and the caller should decide what to do.

File: avocado/core/jobdata.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    if matches == 1:
+        return os.path.dirname(match_file)
+    else:
+        return None
+
+
+def get_id(path, jobid):
+    """
+    Gets the full Job ID using the results directory path and a partial
+    Job ID or the string 'latest'.
```
Comment (ldoktor) [inline avocado/core/jobdata.py]: I have to say I quite dislike this function. It's more-like `check_path_and_id` function and it's being used for 2 different purposes in the code. Anyway that's for some later refactoring.

File: avocado/plugins/replay.py [hunk 8]
```diff
@@ -9,19 +9,22 @@
 #
 # See LICENSE for more details.
 #
-# Copyright: Red Hat Inc. 2013-2014
-# Author: Lucas Meneghel Rodrigues <lmr@redhat.com>
+# Copyright: Red Hat Inc. 2016
+# Author: Amador Pahim <apahim@redhat.com>
```
Comment (ldoktor) [inline avocado/plugins/replay.py]: This should be a different commit with a proper explanation (which is I guess copy&paste issue as the file did not exists in 2014)

File: avocado/plugins/replay.py [hunk 9]
```diff
@@ -88,9 +91,26 @@ def _valid_ignore(self, string):
...
 
     def load_config(self, resultsdir):
-        config = os.path.join(resultsdir, 'replay', 'config')
-        with open(config, 'r') as f:
-            settings.process_config_path(f.read())
+        config = jobdata.retrieve_config(resultsdir)
+        if config is not None:
+            settings.process_config_path(config)
+
+    def _create_replay_map(self, resultsdir, replay_filter):
```
Comment (ldoktor) [inline avocado/plugins/replay.py]: This could use some docstring. First I thought it should be part of the `jobdata`, but after studying the purpose a bit more it really belongs only here.

File: selftests/functional/test_replay_basic.py [hunk 10]
```diff
@@ -58,9 +58,9 @@ def test_run_replay_latest(self):
         self.run_and_check(cmd_line, expected_rc)
 
     def test_run_replay_data(self):
-        file_list = ['multiplex', 'config', 'urls', 'pwd', 'args']
+        file_list = ['multiplex', 'config', 'urls', 'pwd', 'args', 'cmdline']
         for filename in file_list:
-            path = os.path.join(self.jobdir, 'replay', filename)
+            path = os.path.join(self.jobdir, 'jobdata', filename)
```
Comment (ldoktor) [inline selftests/functional/test_replay_basic.py]: OK, but this adds a fallback feature, which ought to be tested. I'd suggest copying at least a simple replay check, run the job, renamed the `jobdata` to `replay` and then run the test on that directory.

Comment (ldoktor) [inline selftests/functional/test_replay_basic.py]: One more thing, I don't see the results-dir documentation and as a matter of fact, it's not documented. Would you please include `jobdata` in the `ReferenceGuide`?

File: avocado/core/jobdata.py [hunk 11]
```diff
@@ -170,7 +170,11 @@ def get_resultsdir(logdir, jobid):
     """
     Gets the job results directory using a Job ID.
     """
-    if jobid == 'latest':
+    if os.path.isdir(jobid):
+        return os.path.expanduser(jobid)
+    elif os.path.isfile(jobid):
```
Comment (ldoktor) [inline avocado/core/jobdata.py]: Hmm, do we want this? (never noticed we can pass a file and use it's parent as results dir)

File: avocado/core/jobdata.py [hunk 12]
```diff
@@ -203,7 +207,9 @@ def get_id(path, jobid):
     Gets the full Job ID using the results directory path and a partial
     Job ID or the string 'latest'.
     """
-    if jobid == 'latest':
+    if os.path.isdir(jobid) or os.path.isfile(jobid):
+        jobid = ''
```
Comment (ldoktor) [inline avocado/core/jobdata.py]: I find this way hackish, but I already mentioned this in the previous commit... (could be fixed later)

File: avocado/plugins/diff.py [hunk 13]
```diff
@@ -329,17 +329,10 @@ def _get_job_data(jobdir):
...
-        elif os.path.isfile(job_id):
-            resultsdir = os.path.dirname(os.path.expanduser(job_id))
-            job_id = ''
-        else:
-            logs_dir = settings.get_value('datadir.paths', 'logs_dir',
-                                          default=None)
-            logdir = os.path.expanduser(logs_dir)
-            resultsdir = jobdata.get_resultsdir(logdir, job_id)
+        logs_dir = settings.get_value('datadir.paths', 'logs_dir',
+                                      default=None)
```
Comment (ldoktor) [inline avocado/plugins/diff.py]: The default should be `''`, `None` is going to raise exception in `os.path` operations.

Comment (ldoktor) [inline avocado/plugins/diff.py]: or no default at all...

File: avocado/plugins/diff.py [hunk 14]
```diff
@@ -329,17 +329,10 @@ def _get_job_data(jobdir):
...
+        logdir = os.path.expanduser(logs_dir)
```
Comment (ldoktor) [inline avocado/plugins/diff.py]: This is actually not needed, you use `expanduser` later in the get_resultsdir`. The `expanduser` is quite useful when dealing with some libraries, but you don't need to use it everywhere...

File: selftests/functional/test_replay_basic.py [hunk 15]
```diff
@@ -57,6 +57,13 @@ def test_run_replay_latest(self):
         expected_rc = exit_codes.AVOCADO_ALL_OK
         self.run_and_check(cmd_line, expected_rc)
 
+    def test_run_replay_jobdir(self):
```
Comment (ldoktor) [inline selftests/functional/test_replay_basic.py]: Some dosctrings could help in this (well, everywhere in our selftests). It took me a while to see the difference.

**Comment by ldoktor** [comment]
Hi @apahim, the unification is very nice, there are few issues here and there, some of them not related to this PR so feel free to ignore them or create cards if you find them useful.

It'd be nice if you could split the patch even more, but I understand it's harder said than done so I think I'm able to merge it even like this, it's just harder for review.

File: avocado/core/jobdata.py (hunk 11 above)
Comment (ldoktor) [inline avocado/core/jobdata.py]: Actually if I were to specify path to a file, I'd expect it to be the `id` file. But in the end I'd rather remove the support for this as it adds the possibility of specifying local file instead of hash.

Comment (ldoktor) [inline avocado/core/jobdata.py]: (I mean it's quite probable that you have a file called `latest` or very short file in your local directory)

Comment (ldoktor) [inline avocado/core/jobdata.py]: ... this also works for the `latest` directory, that's why I'd prefer keeping the logic as minimal as possible

File: avocado/core/jobdata.py (hunk 2 above)
Comment (apahim) [inline avocado/core/jobdata.py]: they are used more than once... this just keep them organized and try to avoid not updated leftovers.

File: avocado/core/jobdata.py (hunk 3 above)
Comment (apahim) [inline avocado/core/jobdata.py]: There are 3 protocols currently. If you say nothing, it will use the `v0`, which is the original ASCII protocol. The `HIGHEST_PROTOCOL` is the `v2`, the same since python 2.3. As far as you use at least python 2.3 in all machines, this is not expected to have compatibility issues. New versions of pickle are not expected to show up in the python versions we support currently.

File: avocado/core/jobdata.py (hunk 4 above)
Comment (apahim) [inline avocado/core/jobdata.py]: Sure.

File: avocado/core/jobdata.py (hunk 7 above)
Comment (apahim) [inline avocado/core/jobdata.py]: Yes, this is a good candidate for improvements. I will do that in a separate PR afterwards.

File: avocado/core/jobdata.py (hunk 2 above)
Comment (ldoktor) [inline avocado/core/jobdata.py]: OK, no problem. Maybe you can define them as private, though...

File: avocado/core/jobdata.py (hunk 3 above)
Comment (ldoktor) [inline avocado/core/jobdata.py]: No problem, I was just wondering, whether the v0 would be sufficient as we store just very basic data.
